verilog/analysis/checkers/mismatched_labels_rule_test.cc

```diff
@@ -0,0 +1,194 @@
+      {"module foo;\n"
+       "  initial\n"
+       "    begin : foo\n"
+       "    end : ", {SymbolIdentifier, "bar"}, "\n"
+       "endmodule"},
+      // end label missing
+      {"module foo;\n"
+       "  initial\n"
+       "    begin : foo\n"
+       "    end\n"
+       "endmodule"},
+      // begin label missing
+      {"module foo;\n"
+       "  initial\n"
+       "    ", {TK_begin, "begin"}, "\n"
+       "    end : foo\n"
+       "endmodule"},
+
+      // for block
+      // no labels
+      {"module foo;\n"
+       "  initial for(int i=0; i<5; ++i)\n"
+       "    begin\n"
+       "    end\n"
+       "endmodule"},
+      // both labels are correct
+      {"module foo;\n"
+       "  initial for(int i=0; i<5; ++i)\n"
+       "    begin : foo\n"
+       "    end : foo\n"
+       "endmodule"},
+      // end label mismatch
+      {"module foo;\n"
+       "  initial for(int i=0; i<5; ++i)\n"
+       "    begin : foo\n"
+       "    end : ", {SymbolIdentifier, "bar"}, "\n"
+       "endmodule"},
+      // end label missing
+      {"module foo;\n"
+       "  initial for(int i=0; i<5; ++i)\n"
+       "    begin : foo\n"
+       "    end\n"
+       "endmodule"},
+      // begin label missing
+      {"module foo;\n"
+       "  initial for(int i=0; i<5; ++i)\n"
+       "    ", {TK_begin, "begin"}, "\n"
+       "    end : foo\n"
+       "endmodule"},
+
+      // nested blocks
+      // no labels
+      {"module foo;\n"
+       "  initial begin\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin\n"
+       "    end\n"
+       "  end\n"
+       "endmodule"},
+      // all labels are correct
+      {"module foo;\n"
+       "  initial begin : first_label\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin : second_label\n"
+       "    end : second_label\n"
+       "  end : first_label\n"
+       "endmodule"},
+      // end label mismatch
+      {"module foo;\n"
+       "  initial begin : first_label\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin : second_label\n"
+       "    end : ", {SymbolIdentifier, "inv_second_label"}, "\n"
+       "  end : first_label\n"
+       "endmodule"},
+      {"module foo;\n"
+       "  initial begin : first_label\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin : second_label\n"
+       "    end : second_label\n"
+       "  end : ", {SymbolIdentifier, "inv_first_label"}, "\n"
+       "endmodule"},
+      {"module foo;\n"
+       "  initial begin : first_label\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin : second_label\n"
+       "    end : ", {SymbolIdentifier, "inv_second_label"}, "\n"
+       "  end : ", {SymbolIdentifier, "inv_first_label"}, "\n"
+       "endmodule"},
+      {"module foo;\n"
+       "  initial begin : first_label\n"
+       "    for(int i=0; i<5; ++i)\n"
+       "    begin : second_label\n"
+       "    end : ", {SymbolIdentifier, "first_label"}, "\n"
+       "  end : ", {SymbolIdentifier, "second_label"}, "\n"
+       "endmodule"},
```

Comment:
These two tests fail with the current approach. The reason is currently the lint rule matches the begin nodes and then tries to find the corresponding end nodes. In that case, when the error is reported on the end node label, the violations are reported in the wrong order (that is, in the order of the begin nodes, not in the order of the end nodes).

I have created this draft PR early, so most of the code can be already reviewed, but this still has to be addressed.

Reply:
These test cases look great.

One thing we can do to address violation ordering is to update the RunLintTestCases driver to sort the violations by starting position.  

It may even be a good idea to sort the violations some time before they are eventually reported to stdout; it is reasonable for users to expect violations to be reported in location order.  Multiple ways this could be done, e.g. overrideable virtual function hook in LintRule to perform arbitrary actions on-finish (which would address unit-testing), or collect all violations after all checks are run and do a one-time final sort before printing.

Reply:
Addressed in #91.